### PR TITLE
Fixed #32233 -- Cleaned-up duplicate connection functionality.

### PR DESCRIPTION
**Pull request by apollo13**

### CODE REVIEW

**Comment by pope1ni** [comment]
Interesting idea. This currently handles database and cache connections. Any thought about using the same interface for email backend connections? (https://docs.djangoproject.com/en/1.11/topics/email/#topic-email-backends)

**Comment by apollo13** [comment]
The thing here is that smtp connections are generally not kept open. But yeah, I'd like to abstract that further

**Comment by apollo13** [comment]
Hi @carltongibson,

sure; it probably just requires a bit of thinking if there could be more cleanups etc…

**Comment by auvipy** [comment]
this would be good to include after rebase.

**Comment by apollo13** [comment]
@auvipy Please stop such comments, they add unnecessary notifications for people watching those issues and do not add any value. Instead of writing such a comment it would be more helpful to fix what you think are the remaining issues and submit a new PR.

**Comment by apollo13** [comment]
@felixxm As discussed somewhen the last weeks I did rebase it and it still seems to pass nicely. I'd appreciate another set of eyes and then the question is if we merge it as is and build upon it (Stuff like connection close semantics could be implemented easily for all backends etc…)

File: django/core/cache/__init__.py [hunk 1]
```diff
@@ -29,88 +27,26 @@
...
-            conf = settings.CACHES[backend]
-        except KeyError:
-            try:
-                # Trying to import the given backend, in case it's a dotted path
-                import_string(backend)
-            except ImportError as e:
-                raise InvalidCacheBackendError("Could not find backend '%s': %s" % (
-                    backend, e))
-            location = kwargs.pop('LOCATION', '')
-            params = kwargs
```
Comment (felixxm) [inline django/core/cache/__init__.py:45]: I'm not sure about removing this. Can we safely remove it?

**Review by felixxm — COMMENTED** [review]
@apollo13 Thanks :+1: I pushed small edits and added one test.

I've also noticed crash when running tests on Oracle:
```
Testing against Django installed in '/django/django' with up to 4 processes
Creating test database for alias 'default'...
Traceback (most recent call last):
  File "./runtests.py", line 589, in <module>
    options.timing,
  File "./runtests.py", line 320, in django_tests
    extra_tests=extra_tests,
  File "/django/django/test/runner.py", line 723, in run_tests
    old_config = self.setup_databases(aliases=databases)
  File "/django/django/test/runner.py", line 643, in setup_databases
    debug_sql=self.debug_sql, parallel=self.parallel, **kwargs
  File "/django/django/test/utils.py", line 183, in setup_databases
    serialize=connection.settings_dict['TEST'].get('SERIALIZE', True),
  File "/django/django/db/backends/base/creation.py", line 55, in create_test_db
    self._create_test_db(verbosity, autoclobber, keepdb)
  File "/django/django/db/backends/oracle/creation.py", line 32, in _create_test_db
    with self._maindb_connection.cursor() as cursor:
  File "/django/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/django/django/db/backends/oracle/creation.py", line 28, in _maindb_connection
    return DatabaseWrapper(settings_dict, alias=self.connection.alias)
  File "/django/django/db/backends/oracle/base.py", line 218, in __init__
    use_returning_into = self.settings_dict["OPTIONS"].get('use_returning_into', True)
KeyError: 'OPTIONS'
```
EDIT: Removing `deepcopy()` fixes this issue.

File: django/core/cache/__init__.py (hunk 1 above)
Comment (felixxm) [reply django/core/cache/__init__.py:45]: ... and if can, we should do this in advance.

**Comment by pope1ni** [comment]
> EDIT: Removing `deepcopy()` fixes this issue.

Yep. I can see the thought process behind doing the `deepcopy`, but it is backward incompatible with assumptions made in various places that entries in `settings.DATABASES` (and possibly `settings.CACHES`) can be modified. If we don't want to allow that we'll need to fix up the places that do this and have some sort of deprecation period, if possible. Regardless, it's a separate issue.

File: django/db/utils.py [hunk 2]
```diff
@@ -131,39 +132,34 @@ def load_backend(backend_name):
...
+            raise ImproperlyConfigured(
+                f"You must define a '{DEFAULT_DB_ALIAS}' database."
+            )
+        if databases[DEFAULT_DB_ALIAS] == {}:
+            databases[DEFAULT_DB_ALIAS]['ENGINE'] = 'django.db.backends.dummy'
+        return databases
+
+    @cached_property
+    def databases(self):
+        return self.settings
```
Comment (pope1ni) [inline django/db/utils.py:162]: `self.settings` is already a `@cached_property`. If the cache needed clearing we'd need to then clear both `self.settings` and `self.databases`. So perhaps we should use `@property` here instead?

Anyway, ISTM that this is not something that is documented and we could just do away with `.databases` entirely and use `.settings` instead?

File: django/utils/connection.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+class ConnectionDoesNotExist(Exception):
+    pass
+
+
+class BaseConnectionHandler:
+    settings_name = None
+    exception_class = ConnectionDoesNotExist
+    thread_critical = False
+
```
Comment (pope1ni) [inline django/utils/connection.py:40]: Again, I suspect we could use `__slots__` here:

```suggestion
    __slots__ = ('_connections', '_settings')

```

File: django/utils/connection.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        self._settings = self.configure_settings(self._settings)
+        return self._settings
+
+    def configure_settings(self, settings):
+        if settings is None:
+            settings = deepcopy(getattr(django_settings, self.settings_name))
+        return settings
+
+    def create_connection(self, alias):
+        raise NotImplementedError('Subclasses must implement create_connection()')
```
Comment (pope1ni) [inline django/utils/connection.py:56]:
```suggestion
        raise NotImplementedError('Subclasses must implement create_connection().')
```

File: tests/utils_tests/test_connection.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from django.test import SimpleTestCase
+from django.utils.connection import BaseConnectionHandler
+
+
+class BaseConnectionHandlerTests(SimpleTestCase):
+    def test_create_connection(self):
+        handler = BaseConnectionHandler()
+        msg = 'Subclasses must implement create_connection()'
```
Comment (pope1ni) [inline tests/utils_tests/test_connection.py:8]:
```suggestion
        msg = 'Subclasses must implement create_connection().'
```

File: django/db/utils.py [hunk 6]
```diff
@@ -173,7 +169,7 @@ def ensure_defaults(self, alias):
         try:
             conn = self.databases[alias]
         except KeyError:
-            raise ConnectionDoesNotExist("The connection %s doesn't exist" % alias)
+            raise self.exception_class(f"The connection {alias} doesn't exist")
```
Comment (pope1ni) [inline django/db/utils.py:172]:
```suggestion
            raise self.exception_class(f'The connection {alias!r} doesn't exist.')
```

File: django/db/utils.py [hunk 7]
```diff
@@ -193,7 +189,7 @@ def prepare_test_settings(self, alias):
...
```
Comment (pope1ni) [inline django/db/utils.py:192]:
```suggestion
            raise self.exception_class(f'The connection {alias!r} doesn't exist.')
```

File: django/utils/connection.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __getitem__(self, alias):
+        try:
+            return getattr(self._connections, alias)
+        except AttributeError:
+            if alias not in self.settings:
+                raise self.exception_class(f"The connection '{alias}' doesn't exist")
```
Comment (pope1ni) [inline django/utils/connection.py:63]:
```suggestion
                raise self.exception_class(f'The connection {alias!r} doesn't exist.')
```

**Review by pope1ni — COMMENTED** [review]
@apollo13 I'm generally happy with this and think it is a beautiful clean up. Thank you!

Agreed that we should try to get this landed before seeing if we can abstract out the connection close stuff, etc.

File: django/core/cache/__init__.py (hunk 1 above)
Comment (felixxm) [reply django/core/cache/__init__.py:45]: Pushed to a separate PR #13753.

**Comment by felixxm** [comment]
I rebased from master, removed `deepcopy()`, and added tests for nonexistent aliases.

File: django/db/utils.py (hunk 2 above)
Comment (felixxm) [reply django/db/utils.py:162]: I think we should leave `databases` for API stability, even if it's not documented.

File: django/utils/connection.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+from django.conf import settings as django_settings
+from django.utils.functional import cached_property
+
+
+class ConnectionProxy:
+    """Proxy for accessing a connection object's attributes."""
+
+    def __init__(self, connections, alias):
+        self.__dict__['_connections'] = connections
+        self.__dict__['_alias'] = alias
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/utils/connection.py:12]: Handlers are not frequently created and can be monkey-patched so I don't think we shouldn't use  `__slots__` here, see ticket-12826.

File: django/utils/connection.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        return getattr(self._connections[self._alias], item)
+
+    def __setattr__(self, name, value):
+        return setattr(self._connections[self._alias], name, value)
+
+    def __delattr__(self, name):
+        return delattr(self._connections[self._alias], name)
+
+    def __contains__(self, key):
+        return key in self._connections[self._alias]
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/utils/connection.py:24]: I think it's fine to leave it.

**Comment by felixxm** [comment]
I moved tests to separate commits and pushed final edits.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/utils/connection.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/utils/connection.py:12]: That's fair, the number of these created should be quite limited.

File: django/db/utils.py (hunk 2 above)
Comment (pope1ni) [reply django/db/utils.py]: Ok. 👍

**Comment by apollo13** [comment]
@pope1ni & @felixxm Uff, well done; anything you need me to do here or will you finish it? :)

**Comment by pope1ni** [comment]
@apollo13 nothing left from me.

**Pull request merged**